haystack_experimental/core/pipeline/pipeline.py

```diff
@@ -45,10 +57,37 @@ def _run_component(
             component_name=component_name, component=component, inputs=inputs
         )
 
+        # NOTE: a workaround for the DocumentJoiner and BranchJoiner components, since _consume_component_inputs()
+        # wraps 'documents' in an extra list, so if there's a 3 level deep list, we need to flatten it to 2 levels only
+        # ToDo: investigate why this is needed and if we can remove it
```

Comment:
I don't know a solution off the top of my head but this is related to our Variadic type. Specifically `Variadic` is a subclass of `Iterable`. So in all of our joiners we have type `Variadic[SomeType]` which basically translates to `Iterable[SomeType]`. And for the DocJoiner this becomes `Iterable[List[Document]]` hence why we get an extra wrapped list. 

My guess we will need to use the checks `socket.is_variadic`, `socket.is_greedy` and `is_socket_lazy_variadic` to help solve this case generally.

Reply:
When I bumped into this issue, I noticed that the function that runs before: `_consume_component_inputs()` adds that extra list and it's only for these two joiners Document and Branch (i.e.: the `ListJoiner` and `StringJoiner` are fine)

Reply:
Huh how very odd. Why would ListJoiner and StringJoiner be treated differently here?

Reply:
I'm still investigating this and trying out another approach. Anyway, dC could already test this first version of the breakpoints even with this ugly fix.